File: A2Z_striver/patterns/pattern_12/pattern_12.py
import sys


class Solution:
    """
    Problem: Print the Number Crown.
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Three Loops (Stateless/Mathematical) - YOUR CURRENT CODE
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Pure logic, no mutable state variable 'k'.
    def pattern12_stateless(self, n):
        for i in range(n):
            # 1. Left Numbers (j+1)
            for j in range(i + 1):
                print(j + 1, end="")
            
            # 2. Spaces
            for _ in range(2 * (n - i - 1)):
                print(" ", end="")
            
            # 3. Right Numbers (Stateless Math)
            # Logic: We want to print (i+1) down to 1.
            # j goes 0 to i.
            # Formula: (i + 1) - j
            for j in range(i + 1):
                print(i + 1 - j, end="")
            
            print()

    # // ~ Approach 2: Three Loops (Stateful)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Intuitive if you prefer tracking values manually.
    def pattern12_stateful(self, n):
        for i in range(n):
            k = 0
            # Left
            for j in range(i + 1):
                k += 1
                print(k, end="")
            # Spaces
            for _ in range(2 * (n - i - 1)):
                print(" ", end="")
            # Right (Reuse k)
            for j in range(i + 1):
                print(k, end="")
                k -= 1
            print()

    # // ~ Approach 3: Single Loop (Coordinate Logic)
    # // Analysis: Time O(N^2) | Space O(1)
    # // Pros: Compact, tests index mastery.
    def pattern12_single_loop(self, n):
        for i in range(n):
            k = 0
            for j in range(2 * n):
                if j <= i:
                    k += 1
                    print(k, end="")
                elif j >= 2 * n - 1 - i:
                    print(k, end="")
                    k -= 1
                else:
                    print(" ", end="")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 4
    print(f"--- Approach 1: Stateless (n={n}) ---")
    solver.pattern12_stateless(n)

File: A2Z_striver/patterns/pattern_12/pattern_12.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print the Number Crown (Double Triangle).
// Example Input: n = 4
// Example Output:
// 1      1
// 12    21
// 123  321
// 12344321
// Goal: O(N^2) Time.

class Solution
{
public:
    // ~ Approach 1: Single Inner Loop (Complex Math Logic)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Compact code, shows strong index manipulation skills.
    void pattern12_single_loop(int n)
    {
        for (int i = 0; i < n; i++)
        {
            int k = 0;
            // Iterate across the full width of the row
            // Width = 2 * n (e.g., if n=4, width=8, indices 0 to 7)
            for (int j = 0; j < 2 * n; j++)
            {

                // --- Logic Derivation ---
                // 1. Left Triangle:
                //    - We need numbers in the first (i+1) columns.
                //    - Indices: 0, 1, ... i
                //    - Condition: j <= i

                if (j <= i)
                {
                    k++;
                    cout << k;
                }

                // 2. Right Triangle:
                //    - This is a mirror of the left.
                //    - Total Width Indices: 0 to (2*n - 1).
                //    - If Left Triangle ends at index 'i', Right Triangle must start 'i' steps from the end.
                //    - Last Index = 2*n - 1
                //    - Start Threshold = (Last Index) - i
                //    - Condition: j >= 2*n - 1 - i
                else if (j >= 2 * n - 1 - i)
                {
                    cout << k;
                    k--; // Decrement as we move right
                }
                else
                {
                    cout << " ";
                }
            }
            cout << "\n";
        }
    }

    // ~ Approach 2: Three Separate Loops (Standard/Readable)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Decoupled logic, easy to debug.
    void pattern12_standard(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // 1. Print Left Numbers (1 to i+1)
            for (int j = 1; j <= i + 1; j++)
            {
                cout << j;
            }

            // 2. Print Spaces
            // Formula: Total Width (2*n) - Left# - Right#
            // Spaces = 2*n - 2*(i+1)
            int spaces = 2 * n - 2 * (i + 1);
            for (int j = 0; j < spaces; j++)
            {
                cout << " ";
            }

            // 3. Print Right Numbers (i+1 down to 1)
            for (int j = i + 1; j >= 1; j--)
            {
                cout << j;
            }

            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 4;

    cout << "--- Approach 1: Single Loop (Math Logic) ---\n";
    sol.pattern12_single_loop(n);

    cout << "\n--- Approach 2: Standard Split Loops ---\n";
    sol.pattern12_standard(n);

    return 0;
}